addons/hr_holidays/models/hr_leave_type.py

```diff
@@ -212,7 +214,116 @@ def _search_virtual_remaining_leaves(self, operator, value):
+                    if allocation_employees[employee_id][holiday_status_id]:
+                        allocations = allocation_employees[employee_id][holiday_status_id] & leaves_employees[employee_id][holiday_status_id]
+                        available_allocations = self.env['hr.leave.allocation']
+                        for allocation_interval in allocations._items:
+                            available_allocations |= allocation_interval[2]
+                        # Consume the allocations that are close to expiration first
+                        available_allocations = available_allocations.sorted(key='date_to')
+                        leave_intervals = leaves_employees[employee_id][holiday_status_id]._items
+                        for leave_interval in leave_intervals:
+                            leave = leave_interval[2]
+                            if leave.leave_type_request_unit == 'day':
+                                leave_duration = leave.number_of_days
+                                leave_unit = 'days'
+                            else:
+                                leave_duration = leave.number_of_hours_display
+                                leave_unit = 'hours'
+                            for available_allocation in available_allocations:
+                                virtual_remaining_leaves = (available_allocation.number_of_days if leave_unit == 'days' else available_allocation.number_of_hours_display) - allocations_days_consumed[employee_id][holiday_status_id][available_allocation]['virtual_leaves_taken']
+                                max_leaves = min(virtual_remaining_leaves, leave_duration)
+                                allocations_days_consumed[employee_id][holiday_status_id][available_allocation]['virtual_leaves_taken'] += max_leaves
+                                if leave.state == 'validate':
+                                    allocations_days_consumed[employee_id][holiday_status_id][available_allocation]['leaves_taken'] += max_leaves
+                                leave_duration -= max_leaves
+                            if leave_duration > 0:
+                                # There are not enough allocation for the number of leaves
+                                allocations_days_consumed[employee_id][holiday_status_id][False]['virtual_remaining_leaves'] -= leave_duration
+                                return allocations_days_consumed
+
+        # Future available leaves
+        for employee_id in allocation_employees:
+            for holiday_status_id in allocation_employees[employee_id]:
+                if allocation_employees[employee_id][holiday_status_id]:
+                    future_allocation_intervals = allocation_employees[employee_id][holiday_status_id] & Intervals([
+                        (fields.datetime.combine(date, time.min), fields.datetime.combine(date, time.max) + timedelta(days=100*365), self.env['hr.leave'])])
```

Comment:
Maybe take the max leaves date_to here also - 100 years is much more than the life expectancy of humans :-D

Reply:
The max leaves date_to would restrict the forecast of number of days available in the future aka the remaining leaves of the allocations. 
For example, if the employee's last leave was yesterday, the intersection of the two intervals will only return the allocations until that date and no future allocation.
I pushed a new commit to consider the number of days available as infinite when the allocation has an end date equal to `fields.datetime.combine(date, time.max) + timedelta(days=100*365)` to avoid using `_get_work_days_data_batch` over an entire century every time :sweat_smile: